### DISCUSSION TITLE
Proposal: inline scripts

### DISCUSSION THREAD

**Discussion by shrinktofit** [cocos/cocos-engine#12349] [Ideas]
How much works you **have to** do if you want to have a button which can toggle a boolean variable?

1. Create a component script in asset manager panel.

2. Double click that script then add a method, fill the button click callback signature. Then code.

3. Go back to editor.

4. Select the button node. Scroll down to to click event array. Click to add an array item.

5. Select the node where the component you've just writtten attached. Usually the button node itself.

6. Select the component...Wait, you should have forgot to attach the component.

7. Now attach the component to the button node.

8. Select the component.

9. Select the method served as the button click event callback.

You can see how much works you should do before everything goes fine.

It would be nice if we could provide a capability to "fast prototyping" if the button click event is not so complex.

### Idea

Each asset may have some scripts inline. In the case of UI node(in scene asset), UI event can be expressed in the form of inline scripts.

The inline scripts are treated as mini-modules like regular file modules. They have URLs same as its owner asset.

Let's re-examine what should we do in the way of inline scripts:

1. Select the button node. Scroll down to to click event array. Click to add an array item. Scroll down to to click event array. Click to add an array item.

2. Code.

That's all things.

----------------------

Update:

⭐ Imaginary effect：

![image](https://user-images.githubusercontent.com/28807867/188058938-f66827d2-bac9-4edf-8b3f-6a67595fcdba.png)

**Comment by jareguo** [comment]
Pretty impeccable advice and manageable maintenance costs.
A similar suggestion for inline animation clips https://github.com/cocos/2d-tasks/issues/110
